### PR TITLE
Fixing binomial

### PR DESCRIPTION
**Pull request by GiovanniVolta**
_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_

## What does the code in this PR do / what does it improve?
Fixing p-value calculation and add the chance to save S1 contributing PMTs for top and bottom (addressing #930 ).

## Can you briefly describe how it works?

Three problems were occurring before. They are listed here below.
1. For one-side left test, namely when the p-value is given by the sf(k;n,p) = 1 - cdf(k;n,p), we were not accounting for the probability of the observation k. Then we were doing p-value = P(x > k) instead of P(x >= k).
2. If j, the point with same probability of the observation k, didn't exist, we were assigning to it 0 or n. Now we assign nan. This allows to write easily the p-value calculation.
3. The interval of the secant method into which look for j were suboptimal. This was leading to wrong j.

The origin of acceptance saw by @skazama at low S1 was due to this.

Last but not least, for the `discrete` test, the scipy one is much more reliable. This would require rounding to the closet integer the k (S1 top) and n (S1). Unfortunately, numba heats me and this is not yet ready - maybe @WenzDaniel knows how to numbafied it.

## Can you give a minimal working example (or illustrate with a figure)?

In the following plots are shown a few self-explanatory examples.

binom_0 95
binom_0 05
binom_0 20
binom_0 40
binom_0 60

_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [x] _Tests to check the (new) code is working as desired._
  - [x] _Does it solve one of the open issues on github?_

### CODE REVIEW

File: straxen/plugins/event_patternfit.py
Comment (kdund) [inline straxen/plugins/event_patternfit.py:536]: ?

Comment (kdund) [inline straxen/plugins/event_patternfit.py:536]: ?

File: straxen/plugins/event_patternfit.py
Comment (kdund) [inline straxen/plugins/event_patternfit.py:487]: why is j_max not just n? are there no cases where one crossing is more than 1.5 times the distance from the mode than the other one?

**Review by kdund — COMMENTED** [review]
Hi, @GiovanniVolta! thanks a lot for the checks. 
I think in general this is good-- see the one test that fails though, is there a way we can check if this is approx OK?

File: straxen/plugins/event_patternfit.py
Comment (GiovanniVolta) [reply straxen/plugins/event_patternfit.py:536]: I wanted to implement the discrete binomial test implemented in scipy instead of the one there is right now, but numba does like it

Comment (kdund) [reply straxen/plugins/event_patternfit.py:536]: gotcha, lets have that wait the next round then

File: straxen/plugins/event_patternfit.py
Comment (kdund) [reply straxen/plugins/event_patternfit.py:490]: (come to think of it if k == mode, you should just return 1. to begin with, I guess)

File: straxen/plugins/event_patternfit.py
Comment (GiovanniVolta) [reply straxen/plugins/event_patternfit.py:487]: Approaching high S1, the secant method has hard time to find the right j if we ask it to look between [0, mode] or [mode, n] since the interval is very big. Generally the relation:

- for k < mode: P(mode; n, p) < P(k; n, p) < P(mode + 1.5 * distance; n, p) 
- for k > mode: P(mode - 1.5 * distance; n, p) < P(k; n, p) < P(mode; n, p) 

is not true. Anyway, it is reasonably true with a tolerance of 1e-6. However, we could relax it a little, for instance to 2.
The only problem with finding j is when we are extremely close to the mode - especially for high S1, above 1e4. Here, no matter the range, the tolerance is worst. But I don't think it is a problem since we are far from the "bad" p-values.

Here a snip code for testing it with the current binomial test:
```
def valide_distance(f, n, p, step=0.001):
    print(f'n: {n:.6f}\tp: {p:.6f}')
    mode = lbinom_pmf_mode(0, n, 0, (n, p))
    print(f'mode: {mode}')
    k_check = np.arange(0, n, step)
    for k in tqdm.tqdm(k_check):
        target = lbinom_pmf(k, n, p)
        
        distance = abs(mode - k)
        if k != 0 and k != n and binom_pmf(k, n, p) > 1e-06:
            if k < mode: 
                    j_min = mode
                    j_max = min(mode + f * distance, n)
                    j = lbinom_pmf_inverse(j_min, j_max, target, (n, p))
                    if not np.isnan(j) and j != n:
                        if not binom_pmf(j_min, n, p) > binom_pmf(k, n, p) > binom_pmf(j_max, n, p):
                            print(j_min, k, j_max)
                            print(binom_pmf(j_min, n, p), binom_pmf(k, n, p), binom_pmf(j_max, n, p))
                            break
                        if not np.isclose(binom_pmf(j, n, p), binom_pmf(k, n, p), rtol=0, atol=1e-06, equal_nan=False):
                            print(j, k)
                            print(binom_pmf(j, n, p), binom_pmf(k, n, p))
                            break
            elif k > mode:
                    j_min = max(mode - f * distance, 0)
                    j_max = mode
                    j = lbinom_pmf_inverse(j_min, j_max, target, (n, p))
                    if not np.isnan(j) and j != n:
                        if not binom_pmf(j_min, n, p) < binom_pmf(k, n, p) < binom_pmf(j_max, n, p):
                            print(j_min, k, j_max)
                            print(binom_pmf(j_min, n, p), binom_pmf(k, n, p), binom_pmf(j_max, n, p))
                            break
                        if not np.isclose(binom_pmf(j, n, p), binom_pmf(k, n, p), rtol=0, atol=1e-06, equal_nan=False):
                            print(j, k)
                            print(binom_pmf(j, n, p), binom_pmf(k, n, p))
                            break
            else:
                pass
```

File: straxen/plugins/event_patternfit.py
Comment (GiovanniVolta) [reply straxen/plugins/event_patternfit.py:490]: Yes it is intended. The p-vlaue is 1 and if `ls, rs = k, k`. So either I can leave it like this or I can just `return 1`

File: straxen/plugins/event_patternfit.py
Comment (GiovanniVolta) [reply straxen/plugins/event_patternfit.py:474]: Yeah, the precision depends on S1. For low S1 we have high tolerance, but this decrease with the increasing of the area. Can we maybe make it as a warning instead of error ? I will then be very clear in the summary note.

File: straxen/plugins/event_patternfit.py
Comment (kdund) [reply straxen/plugins/event_patternfit.py:487]: For that high S1s, the gaussian approximation is already excellent-- should we use that, perhaps?

File: straxen/plugins/event_patternfit.py
Comment (kdund) [reply straxen/plugins/event_patternfit.py:490]: I think the latter is cleaner for readability

File: straxen/plugins/event_patternfit.py
Comment (kdund) [reply straxen/plugins/event_patternfit.py:474]: Yeah, that sounds good (how to turn test into warning?)

File: straxen/plugins/event_patternfit.py
Comment (GiovanniVolta) [reply straxen/plugins/event_patternfit.py:487]: Yes sure, do you want to me to do it or can we leave it for the next version? I can open an issue to remember that

**Comment by GiovanniVolta** [comment]
@kdund the tests are working now. In the case that k < mode and j doesn't exist, the algorithm set j = n. This doesn't make any problem for the p-value but for the test yes. Now it should be fine. Let me know what you think. 
Cheers!

File: straxen/plugins/event_patternfit.py
Comment (kdund) [reply straxen/plugins/event_patternfit.py]: open issue is fine

**Review by kdund — APPROVED** [review]
Thanks for sticking with it, @GiovanniVolta !

**Comment by GiovanniVolta** [comment]
Do we need another approval?

**Comment by kdund** [comment]
> Do we need another approval?

No, but for straxen I believe we've held off on merging before until we were ready for a new version, or am I misremembering ( @ershockley ?)

**Comment by ershockley** [comment]
Yeah usually we merge PRs that will change lineage close to a new release. This is a high level plugin though so it's probably fine. I leave decision to @WenzDaniel and @JoranAngevaare

**Comment by JoranAngevaare** [comment]
Thanks all, we will merge it when making the new release. We could even consider making one specifically for this one, depending on how urgent this PR is?

>This is a high level plugin though so it's probably fine.

It's indeed high level but one of the computationally more intense plugins

**Pull request merged**
